libbeat/processors/cache/file_store.go

```diff
@@ -287,8 +287,8 @@ func (c *fileStore) writeState(final bool) {
 	enc := json.NewEncoder(f)
 	enc.SetEscapeHTML(false)
 	now := time.Now()
-	for c.expiries.Len() != 0 {
-		e := c.expiries.pop()
+	for i := 0; i < c.expiries.Len(); i++ {
+		e := c.expiries[i]
```

Comment:
```suggestion
	for _, e := range c.expiries {
```
The previously logic was obviously written when we were only planning to write out on exit, so the iterator was destructive. It would be nice if the write out also dropped expired elements, but that cannot happen non-destructively within this iterator. It could be done by a mark during this iteration, followed by a sweep in the postamble.

Not for now, but this would look like
```diff
diff --git a/libbeat/processors/cache/file_store.go b/libbeat/processors/cache/file_store.go
index d3820600ac..062a64a916 100644
--- a/libbeat/processors/cache/file_store.go
+++ b/libbeat/processors/cache/file_store.go
@@ -287,10 +287,12 @@ func (c *fileStore) writeState(final bool) {
        enc := json.NewEncoder(f)
        enc.SetEscapeHTML(false)
        now := time.Now()
-       for c.expiries.Len() != 0 {
-               e := c.expiries.pop()
+       var expired int
+       for _, e := range c.expiries {
                if e.Expires.Before(now) {
-                       // Don't write expired elements.
+                       // Don't write expired elements; mark and continue.
+                       e.Expires = time.Time{}
+                       expired++
                        continue
                }
                err = enc.Encode(e)
@@ -299,6 +301,17 @@ func (c *fileStore) writeState(final bool) {
                        return
                }
        }
+       if expired != 0 {
+               // Fix up heap and sweep cache.
+               heap.Init(&c.expiries)
+               for n := 0; (c.effort <= 0 || n < c.effort) && len(c.cache) != 0; n++ {
+                       if c.expiries[0].Expires.After(now) {
+                               break
+                       }
+                       e := c.expiries.pop()
+                       delete(c.cache, e.Key)
+               }
+       }
        // Only mark as not dirty if we succeeded in the write.
        c.dirty = false
        c.log.Infow("write state to file succeeded", "id", c.id, "path", c.path)
```

Reply:
Not sure this is super worth it, does not this operation of dropping also happen at every Put?

Reply:
PS: not really related, but I noticed the file write can lead to some issues when the cache is really large, since it stops the world until done due to the need of locking the cache. Not sure about how to avoid it, but just pointing it out as it could cause issues if write_interval is on the low end and cache size is on the larger.

Reply:
> Not sure this is super worth it, does not this operation of dropping also happen at every Put?

It does, but only "just enough". This takes a small amount of time relative to the file write to minimize the memory usage.

> PS: not really related, but I noticed the file write can lead to some issues when the cache is really large, since it stops the world until done due to the need of locking the cache. Not sure about how to avoid it, but just pointing it out as it could cause issues if write_interval is on the low end and cache size is on the larger.

If this does become an issue, the approach would be to:
```
try lock write
return if unsuccess
defer unlock write

go:
    lock cache
    obtain worklist of entries more recent than the expiry time
    for each worklist item:
        write element if still more recent than noted time of heap front
    update note of time of current heap front
    unlock cache
    for each remaining worklist item: # or maybe a small set of worklist
        lock cache
        write element if still more recent than expiry time
        unlock cache
```
This requires adding a new field to the `memStore` type that keeps a track of the front of the heap for this purpose, and a lock to `fileStore`.

The approach here will potentially write out deleted items, but should never write out invalid items.

However, it's significantly more complex and unless there is actually an issue, I don't think this is worth it; tuning should be a first option.

Reply:
> It does, but only "just enough". This takes a small amount of time relative to the file write to minimize the memory usage.

The only thing a bit odd to me, unless I got something wrong,  is that this memory benefit only applies if using the file write intervals, while when using just memory backend or file without the write intervals, this would just be skipped and the "just enough" effort would be done. That is why I mention that I am not sure is worth it, unless we want to do it somewhere else to benefit all configurations.

Reply:
Yeah, I'm fine with things how they are; the code looks good to me. The only reason I suggested it is that the most common (and likely recommended use would be file-backed) and the cost of this work is small in comparison to the cost of the file write, so it was a reasonable place to insert the work in a way that is routinely run.